File: CAPTURER/InterfaceUtilisateur/SaisieTuile.vb
''' <summary> permet la saisie d'un pointD au format tuile </summary>
Friend Class SaisieTuile
    Private LimiteSite, StockClipCurseur As Rectangle
    Private Titre As String
    Private ReadOnly SaisieXY As TypeSaisie = TypeSaisie.EntierPositif
    ''' <summary> initialisation de la location et des différents champs de saisie </summary>
    Private Sub SaisieTuile_Load(sender As Object, e As EventArgs)
        Titre = TitreInformation
        StockClipCurseur = Cursor.Clip
        'Attend un point dd en PointGrille
        Dim PointCoordonnees As PointD = CType(Tag, PointD)
        If Not PointCoordonnees.IsEmpty Then
            Dim PointPixels As Point = PointGrilleToPointPixel(PointCoordonnees, Serveur.SiteCarto, Affichage.Echelle)
            Dim PG As New PointG(CapturerSettings.INDICE_ECHELLE, PointPixels)
            Col.Text = PG.IndexTuile.X.ToString("#0")
            Col.Select(Col.Text.Length, 0)
            DecalX.Text = PG.Offset.Width.ToString("#0")
            DecalX.Select(DecalX.Text.Length, 0)
            Row.Text = PG.IndexTuile.Y.ToString("#0")
            Row.Select(Row.Text.Length, 0)
            DecalY.Text = PG.Offset.Height.ToString("#0")
            DecalY.Select(DecalY.Text.Length, 0)
        End If
        LimiteSite = RegionGrilleToRegionPixels(Serveur.Limites, Serveur.SiteCarto, Affichage.Echelle)
        Cursor.Clip = New Rectangle(Location, Size)
    End Sub
    ''' <summary> filtrage des touches admises </summary>
    Private Sub Coord_KeyDown(sender As Object, e As KeyEventArgs)
        e.SuppressKeyPress = SuppressionTouche(SaisieXY, e.KeyCode)
    End Sub
    ''' <summary> filtrage du caractère . pour les champs concernant les secondes </summary>
    Private Sub Coord_KeyPress(sender As Object, e As KeyPressEventArgs)
        Dim C As TextBox = CType(sender, TextBox)
        e.Handled = SuppressionCaractere(SaisieXY, C, e.KeyChar)
    End Sub
    ''' <summary> affiche un message d'erreur </summary>
    Private Shared Sub Erreur(Champ As String)
        MessageInformation = $"Le champ {Champ} doit être égal ou supérieur à 0"
        TitreInformation = "Erreur de saisie"
        AfficherInformation()
    End Sub
    ''' <summary> validation des saisies avec message d'erreur en cas d'erreur de saisie </summary>
    Private Sub SaisieTuile_FormClosing(sender As Object, e As FormClosingEventArgs)
        If DialogResult = DialogResult.OK Then
            'renvoie un point double en DD
            If String.IsNullOrEmpty(Col.Text) Then
                e.Cancel = True
                Erreur("Index Tuile X")
                Col.Focus()
                Exit Sub
            End If
            If String.IsNullOrEmpty(Row.Text) Then
                e.Cancel = True
                Erreur("Index Tuile Y")
                Row.Focus()
                Exit Sub
            End If
            If String.IsNullOrEmpty(DecalX.Text) Then
                e.Cancel = True
                Erreur("Decalage Pixels X")
                DecalX.Focus()
                Exit Sub
            End If
            If String.IsNullOrEmpty(DecalY.Text) Then
                e.Cancel = True
                Erreur("Decalage Pixels Y")
                DecalY.Focus()
                Exit Sub
            End If
            Dim Result As New PointG(0, CInt(Col.Text), CInt(Row.Text), CInt(DecalX.Text), CInt(DecalY.Text))
            'Logic point saisi en dehors des limites
            If FlagLimitesSite AndAlso Not LimiteSite.Contains(Result.Location) Then
                e.Cancel = True
                MessageInformation = $"Le point doit être compris entre :{CrLf}{New PointG(CapturerSettings.INDICE_ECHELLE, LimiteSite.Location).ToString}{CrLf}et {CrLf}" &
                                       New PointG(CapturerSettings.INDICE_ECHELLE, Point.Add(LimiteSite.Location, New Size(LimiteSite.Width, LimiteSite.Height))).ToString
                TitreInformation = "Point Hors limites"
                AfficherInformation()
                Exit Sub
            End If
            'renvoie un point double en mètres
            Tag = PointPixelToPointGrille(Result.Location, Serveur.SiteCarto, Affichage.Echelle)
        End If
        Cursor.Clip = StockClipCurseur
        TitreInformation = Titre
    End Sub
End Class

File: CAPTURER/InterfaceUtilisateur/SaisieTuile.cs
using static FCGP.AfficheInformation;
using static FCGP.CapturerGlobal;
using static FCGP.Commun;
using static FCGP.ConvertirCoordonnees;
using static FCGP.Settings;

namespace FCGP
{
    /// <summary> permet la saisie d'un pointD au format tuile </summary>
    internal partial class SaisieTuile
    {
        private Rectangle LimiteSite, StockClipCurseur;
        private string Titre;
        private readonly TypeSaisie SaisieXY = TypeSaisie.EntierPositif;
        /// <summary> initialisation de la location et des différents champs de saisie </summary>
        private void SaisieTuile_Load(object sender, EventArgs e)
        {
            Titre = TitreInformation;
            StockClipCurseur = Cursor.Clip;
            // Attend un point dd en PointGrille
            PointD PointCoordonnees = (PointD)Tag;
            if (!PointCoordonnees.IsEmpty)
            {
                var PointPixels = PointGrilleToPointPixel(PointCoordonnees, Serveur.SiteCarto, Affichage.Echelle);
                var PG = new PointG(CapturerSettings.INDICE_ECHELLE, PointPixels);
                Col.Text = PG.IndexTuile.X.ToString("#0");
                Col.Select(Col.Text.Length, 0);
                DecalX.Text = PG.Offset.Width.ToString("#0");
                DecalX.Select(DecalX.Text.Length, 0);
                Row.Text = PG.IndexTuile.Y.ToString("#0");
                Row.Select(Row.Text.Length, 0);
                DecalY.Text = PG.Offset.Height.ToString("#0");
                DecalY.Select(DecalY.Text.Length, 0);
            }
            LimiteSite = RegionGrilleToRegionPixels(Serveur.Limites, Serveur.SiteCarto, Affichage.Echelle);
            Cursor.Clip = new Rectangle(Location, Size);
        }
        /// <summary> filtrage des touches admises </summary>
        private void Coord_KeyDown(object sender, KeyEventArgs e)
        {
            e.SuppressKeyPress = SuppressionTouche(SaisieXY, e.KeyCode);
        }
        /// <summary> filtrage du caractère . pour les champs concernant les secondes </summary>
        private void Coord_KeyPress(object sender, KeyPressEventArgs e)
        {
            TextBox C = (TextBox)sender;
            e.Handled = SuppressionCaractere(SaisieXY, C, e.KeyChar);
        }
        /// <summary> affiche un message d'erreur </summary>
        private static void Erreur(string Champ)
        {
            MessageInformation = $"Le champ {Champ} doit être égal ou supérieur à 0";
            TitreInformation = "Erreur de saisie";
            AfficherInformation();
        }
        /// <summary> validation des saisies avec message d'erreur en cas d'erreur de saisie </summary>
        private void SaisieTuile_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (DialogResult == DialogResult.OK)
            {
                // renvoie un point double en DD
                if (string.IsNullOrEmpty(Col.Text))
                {
                    e.Cancel = true;
                    Erreur("Index Tuile X");
                    Col.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(Row.Text))
                {
                    e.Cancel = true;
                    Erreur("Index Tuile Y");
                    Row.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(DecalX.Text))
                {
                    e.Cancel = true;
                    Erreur("Decalage Pixels X");
                    DecalX.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(DecalY.Text))
                {
                    e.Cancel = true;
                    Erreur("Decalage Pixels Y");
                    DecalY.Focus();
                    return;
                }
                var Result = new PointG(0, int.Parse(Col.Text), int.Parse(Row.Text), int.Parse(DecalX.Text), int.Parse(DecalY.Text));
                // Logic point saisi en dehors des limites
                if (FlagLimitesSite && !LimiteSite.Contains(Result.Location))
                {
                    e.Cancel = true;
                    MessageInformation = $"Le point doit être compris entre :{CrLf}{new PointG(CapturerSettings.INDICE_ECHELLE, LimiteSite.Location).ToString()}{CrLf}et {CrLf}" + new PointG(CapturerSettings.INDICE_ECHELLE, Point.Add(LimiteSite.Location, new Size(LimiteSite.Width, LimiteSite.Height))).ToString();
                    TitreInformation = "Point Hors limites";
                    AfficherInformation();
                    return;
                }
                // renvoie un point double en mètres
                Tag = PointPixelToPointGrille(Result.Location, Serveur.SiteCarto, Affichage.Echelle);
            }
            Cursor.Clip = StockClipCurseur;
            TitreInformation = Titre;
        }
    }
}